### DISCUSSION TITLE
Development Insight - Resty v3.0.0

### DISCUSSION THREAD

**Discussion by jeevatkm** [go-resty/resty#720] [Announcements]
This discussion aims to provide high-level insight into v3.0.0 development. Feel free to engage and share ideas and inputs.

## Branching Approach

- [x] Branch v2 will be created and set as the default branch on GitHub.
- [x] Rename branch master to branch main.
- [x] The file go.mod on the main branch will be updated to github.com/go-resty/resty/v3 and go version.
    * The v3 feature development will branch out from the main branch. When ready, PR will be created against the main branch.
    * When v3 is ready, the main branch will become the default branch on GitHub.
 
### Client

- [x] Currently, there are three ways to create clients. I will add additional ways to create a Resty client.
  - [x] One will be fully composable middleware during client creation.
  - [x] Make it easy, WASM dial content
- [x] Add an option to provide all the timeout values during client creation.
- [x] Read-Write Mutex will be added into the client struct to make all the client properties thread-safe
   - [x] All the exported fields will become getter methods, e.g., client.BaseURL => client.BaseURL()
- [x] Improving TLS config handling
- [x] Improving Transport handling

### Middleware

- [ ] ~~Introducing Package middleware with all the existing middleware will be exported, and the sub-function will be exposed as complete middleware.~~
- [x] The middleware approach will be first-class citizens in tackling many practical scenarios, such as authentication, authorization, tracing, content-type handling, debug logger, and much more, in the request and response processing pipeline.
- [x] Pluggable Service discovery

### Retry, Redirect

* ~~Debating about creating the package; it is still being decided!! Is it worth it??~~

### And so on.

- [x] First v3-selected labeled issues will be focused from v3.0.0 Milestone
- [x] Then, the v3-consideration labeled issues will be focused on.
   - [x] Deprecated line items will be removed
- [ ] ~~Introducing package internal to park certain internal functionality~~

## Documentation and Website

- [x] Resty needs to move away from Readme and provide as much how-to documentation as possible. Resty users' contributions are welcome.
- [x] Go lang supports the latest two versions. I'm debating whether two or four of the latest versions result in a minimum version on Resty's go.mod file.
- [x] I have purchased the two TLDs, resty.dev and ~~go-resty.org~~, and I'm wondering which one to use.
   - [x] I'm going with resty.dev domain
- [x] We need to create a simple and effective logo for Resty. I'm looking for input from Resty users. 

--------
So far, I have captured my thoughts and will continue to do so here.

**Comment by SVilgelm** [comment]
Also need to decide if the requests objects are immutable as well as the client objects by the internal code. If it is, then we can reuse the request objects, otherwise make a clear statement in the documentation.

I think the client object should be thread-safe via RW mutex, but the request should not.

**Comment by jeevatkm** [comment]
I'm considering keeping the Request object's nature as-is, and I agree on the client object with RW mutex.

**Comment by pior** [comment]
Is there a target release date for v3? Probably not, of course, I'm just trying to get an idea of the timeline.
The v3 milestone seems almost fully done, but is there a lot of additional work you want to achieve before the release?

**Reply by jeevatkm** [reply]
@pior, thanks for checking in! I'm almost done with version 3, but there are still a few more things I need to add and improve before it’s ready.

I'm aiming for a beta version in the coming January (unless something pops up). The plan is -
- Finish the remaining dev line items
- Create v3 documentation; I'm aiming to do example-driven documentation
- Publish the beta version of v3 and keep it for a couple of months to get feedback and fix any issues that arise
- Then publish the final release of v3

**Comment by jeevatkm** [comment]
Yesterday, v3 first beta version was released - https://github.com/go-resty/resty/releases/tag/v3.0.0-beta.1

**Reply by ekafeel** [reply]
When will it be out of beta? I want to use it in my project, but I'm hesitant because of this.

**Reply by takanuva15** [reply]
I'm interested in the timeline for official release too. Can you clarify what components are left to fix before this moves out of beta? (It looks like all the checkboxes in the main description are already completed)

**Reply by jeevatkm** [reply]
@pior @ekafeel @takanuva15 Major line items are done. There are a few more closing parts left. In the upcoming days, I will be making the final beta version `v3.0.0-beta.7` in a week or so. 

After that, I will work on the Godoc and documentation, which will also give me time to get feedback before the RC1 release.

**Comment by pior** [comment]
@jeevatkm How can we help push the v3 release forward?

**Comment by vadimi** [comment]
hello @jeevatkm, any news/plans on v3 release?

**Reply by jeevatkm** [reply]
@vadimi @pior Thanks for reaching out. I have been making good progress. I'm aiming for v3 RC1 in a few weeks.

**Reply by pior** [reply]
🎉 
Tag me if there's anything I can help with.

**Reply by jeevatkm** [reply]
@pior I think I will be able to make v3 RC1 release this weekend or early next week. Do you mind reviewing docs in this branch [`for-v3.0.0-rc.1-release`](https://github.com/go-resty/docs/tree/for-v3.0.0-rc.1-release)?
Notably, migration guide, new feature and enhancements, etc.

**Reply by pior** [reply]
I did a quick pass: https://github.com/go-resty/docs/pull/20
I don't often work on docs, I don't know if those changes make sense

**Comment by jeevatkm** [comment]
v3 RC1 is released today. Thanks for your support.
